### PR TITLE
Add tables to systems and domains

### PR DESCRIPTION
**Pull request by Fox32**
Hmm, got a bit bigger than expected...

You can now drill down from the domains in the explore plugin down to subcomponents and vice-versa.

This PR adds tables to the domain and system entity page displaying the related entities (and provides a table for subcomponents). Tables are a simple solution for now, but might evolve into a better concept around visualizing relations in the future. 

image

image

image

I also updated the "empty state" style for the tables, as the previous design took to much space and it always looked like an error having no API consumers. This includes the existing tables around APIs:

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [inline plugins/catalog/src/plugin.ts:68]: Could prolly skip the last path component here right? using just `./components/HasSystemsCard` to keep it short :p

**Comment by adamdmharvey** [comment]
Just wanted to say: really nicely done, @Fox32!  Love all these contributions that help better make sense of the data that's in the catalog. Gives us a great point to continue to iterate and elevate understanding of the software in the catalog at scale. Thanks for all you do for the project! 👏 👏 👏 👏 🚀 🤩

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [inline plugins/catalog/src/plugin.ts:64]: Bit hesitant about adding these here tbh, although not sure were they would go instead. This plugin can easily grow out of control xP

**Review by Rugvip — COMMENTED** [review]
Awesome stuff, thanks! :grin:

File: plugins/catalog/src/plugin.ts
Comment (Fox32) [reply plugins/catalog/src/plugin.ts:68]: Would need a index.ts in that folder, but personally would prefer having one. I think I just keept this pattern because the `EntityPage` is doing it above 😆

File: plugins/catalog/src/plugin.ts
Comment (Fox32) [reply plugins/catalog/src/plugin.ts:62]: Hmm, yeah resonable. I'm not sure if these tables are the final solution, but an easy thing we can do for now to get at least a full connection between the entities going.
We could create a separate ecosystem plugin, but on the other hand I'm already questioning why we have separate `org` and `api-docs` plugins.

Comment (Rugvip) [reply plugins/catalog/src/plugin.ts:62]: Yep idd, think we need a bit more info and thinking here before we can make good decisions. No matter what I think we're gonna wanna end up wanting to do some breaking changes here in a bit, so don't wanna block the PR on this discussion

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [reply plugins/catalog/src/plugin.ts]: Ah alright xP

Yeah it's usually best to have an index file that exports only the things that should be usable outside, since sometimes you want to export things for tests that you may not want to be used

File: plugins/catalog/src/plugin.ts
Comment (Fox32) [reply plugins/catalog/src/plugin.ts:62]: @Rugvip How do we continue here?

Comment (Rugvip) [reply plugins/catalog/src/plugin.ts:62]: Like this :grin:

File: plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx
Comment (Rugvip) [inline plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: Tbh I think `SystemsTable` (and component) are too specific for this package. Perhaps move to the catalog plugin for now or smth like that?

I don't think this package should contain any logic related to specific entity kinds, since that's a dimension that is completely open to customization, while this package isn't. Perhaps the best solution here would be a slightly lower level `EntityTable`?

File: .changeset/pink-coins-sniff.md
Comment (Rugvip) [inline .changeset/pink-coins-sniff.md:14]: Bit of a typo here right? 😅

**Review by Rugvip — COMMENTED** [review]
Just some thoughts on the additions to `catalog-react`, otherwise I think we're good to go

File: plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx
Comment (freben) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: On the other hand, does it still make sense to have things here that pertain to the _core_ kinds? It's a pretty well defined small set of kinds and semantics after all. I would like to look at this package as "the toolbox of pieces with which you can glue together your own entity page etc"

Comment (Rugvip) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: As a possible way to quickly make everyone happy, is a single `EntityTable` a reasonable refactor? or does it reduce the value of the component too much

Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: A `EntityTable` would be possible, but the question is, what columns we have in such a table. As far as I see, the only thing all entities have in common are name/entity ref & description (from the parts the tables display right now). So I'm not sure.

File: .changeset/pink-coins-sniff.md
Comment (Fox32) [reply .changeset/pink-coins-sniff.md]: Yep, missed the ComponentsTable and SystemTable...

File: plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx
Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: So maybe to be clear why export these from `catalog-react` at all and not just having the entity cards: I reuse the `ComponentsTable` in `api-docs`. We could move both to `catalog` and not export them and duplicate the `ComponentsTable` into `api-docs`.

Comment (Rugvip) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: Hmm, what do you think about an `EntityTable` + `EntityTable.componentColumns` + `EntityTable.systemColumns` combo? Along with moving over the entity field picking into the column definitions

Provides a clear structure and I think would be simple enough to use, but also space for completely custom solutions.

Something like

```tsx
<EntitiesTable
  title="Consumers"
  variant={variant}
  emptyComponent={ /* bleh */ }
  entities={entities}
  columns={EntitiesTable.componentColumns}
/>
```

Not sure if the columns would be found at `EntitiesTable.*` or not, but tbh they have to exist somewhere, so 😅

Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: I will have a look tomorrow. Just wondering where to do the transformation from entity to row. To make the table work properly I transform the values of the relations before passing them into the table. But these transforms are entity kind specific. However, I may saw a different solution some days ago...

Comment (Rugvip) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: @backstage/maintainers on board with ^ btw? Before @Fox32 puts in work towards that 😅

Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: 😆 already playing with it

Comment (freben) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: `columns: 'domain-defaults' | 'system-defaults' | string[]`? :)

Comment (Rugvip) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: Definitions are a lot more complex than that tho

Comment (freben) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: And we agreed that letting it live in the react package was good, right

Comment (freben) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: Ya right, so something like `() => ColumnWhateverType[]` then (that only gets evaluated on first render) perhaps, but the idea is the same

Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: What do you think about something like this? https://github.com/SDA-SE/backstage/commit/42df711a93e1959d9004442ab88f57d10e41e47f

Comment (Rugvip) [reply plugins/catalog-react/src/components/EntityTables/SystemsTable.tsx:95]: @Fox32 👌

**Comment by Fox32** [comment]
Gave it an update, no visual changes. There is now an `EntityTable` to easily build new tables for different entity kinds, with some presets and helpers.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4392-ttdrrvifexusaijop4kp3jkz6likt6ep.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60264dd0f285afe5898cd639

**Comment by Fox32** [comment]
Not wanne make pressure, but we would like to see this merged. I know it's huge. I can still offer to split it up into small parts.

File: packages/core/src/components/Table/Table.tsx
Comment (Rugvip) [inline packages/core/src/components/Table/Table.tsx:206]: Should prolly be `emptyContent` instead as it isn't a component, could maybe be a `ReactNode` instead of `JSX.Element` as well

**Review by Rugvip — COMMENTED** [review]
Thanks, looks good! 👍  😁 

a minor naming nit

File: plugins/catalog-react/src/components/EntityTable/EntityTable.tsx
Comment (Rugvip) [inline plugins/catalog-react/src/components/EntityTable/EntityTable.tsx:26]:
```suggestion
  variant?: 'gridItem';
```

File: plugins/api-docs/src/components/ApisCards/presets.tsx
Comment (freben) [inline plugins/api-docs/src/components/ApisCards/presets.tsx:32]:
```suggestion
// special createSpecApiTypeColumn. But this is required to use ApiTypeTitle to
```

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx
Comment (freben) [inline plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx:77]:
```suggestion
    expect(getByText(/No component provides this API/i)).toBeInTheDocument();
```

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx
Comment (freben) [inline plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx:86]:
```suggestion
          No component provides this API.{' '}
```

File: plugins/catalog-react/src/components/EntityTable/columns.tsx
Comment (freben) [inline plugins/catalog-react/src/components/EntityTable/columns.tsx:51]:
```suggestion
      // Another alternative would be to cache the values. But writing them
```

File: plugins/catalog-react/src/components/EntityTable/columns.tsx
Comment (freben) [inline plugins/catalog-react/src/components/EntityTable/columns.tsx:52]: Hm we could have a row type that contains contextual data besides the entity itself right? Including "computed" things like a search terms array or whatnot.

File: plugins/catalog-react/src/components/EntityTable/columns.tsx
Comment (freben) [inline plugins/catalog-react/src/components/EntityTable/columns.tsx:53]: Is `filter` the complete string? I guess we should make a helper function that supports tokenisation (space-splitting) of the filter text and make sure that all of those tokens appear in the input

File: plugins/catalog-react/src/hooks/useRelatedEntities.ts
Comment (freben) [inline plugins/catalog-react/src/hooks/useRelatedEntities.ts:30]: No need to use the `Retry` variation if you don't use that facility right? 🤔

File: plugins/catalog-react/src/hooks/useRelatedEntities.ts
Comment (freben) [inline plugins/catalog-react/src/hooks/useRelatedEntities.ts:37]: these could be case insensitive too honestly

File: plugins/catalog/src/components/ComponentsCard/HasComponentsCard.tsx
Comment (freben) [inline plugins/catalog/src/components/ComponentsCard/HasComponentsCard.tsx:56]: yeah i keep thinking the same here as above - why doesn't this one just render `<InfoCard variant={variant}>` and then move all the useEntity and stuff into an inner contents component?

**Review by freben — COMMENTED** [review]
Great stuff! Some questions, but thinking this is ripe for merging very soon

File: plugins/catalog-react/src/components/EntityTable/EntityTable.tsx
Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTable/EntityTable.tsx:26]: I like to use an enum type here. We should fix that for all components that use variant, but I will just do it in the components of this PR for now to keep the changes small. Should also be fixed in the InfoCard itself.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4392-ttdrrvifexusaijop4kp3jkz6likt6ep.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60264dd0f285afe5898cd639

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4392-ttdrrvifexusaijop4kp3jkz6likt6ep.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60264dd0f285afe5898cd639

File: plugins/catalog-react/src/hooks/useRelatedEntities.ts
Comment (Fox32) [reply plugins/catalog-react/src/hooks/useRelatedEntities.ts]: Yep, we don't need that here.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4392-ttdrrvifexusaijop4kp3jkz6likt6ep.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60264dd0f285afe5898cd639

File: plugins/catalog/src/components/ComponentsCard/HasComponentsCard.tsx
Comment (Fox32) [reply plugins/catalog/src/components/ComponentsCard/HasComponentsCard.tsx:56]: I think this is just a left over from a time where I had to repeat more code. I will just use the InfoCard directly.

File: plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx
Comment (Fox32) [reply plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx:81]: Yeah exactly. We either render the table (which behaves as a card) or draw a card on our own. The problem is that the underlining table component from `@backstage/core` behaves like a table. I think there is now way to turn that off. Not sure if this is something we want to solve in this PR.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4392-ttdrrvifexusaijop4kp3jkz6likt6ep.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60264dd0f285afe5898cd639

File: plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx
Comment (freben) [reply plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx:70]: Thanks for explaining this, it was confusing us too. Should definitely break the card and table apart at some point later... Not now.

**Review by freben — APPROVED** [review]

File: plugins/catalog-react/src/components/EntityTable/EntityTable.tsx
Comment (Fox32) [reply plugins/catalog-react/src/components/EntityTable/EntityTable.tsx]: Created #4558

**Pull request merged**
